### PR TITLE
Added support for encrypted pvc

### PR DESCRIPTION
**Pull request by geoffrey1330**

### CODE REVIEW

File: pkg/spdk/controllerserver.go [hunk 1]
```diff
@@ -194,59 +198,69 @@ func (cs *controllerServer) DeleteSnapshot(_ context.Context, req *csi.DeleteSna
...
+	distrNpcs, err := getIntParameter(params, "distr_npcs", 1)
 	if err != nil {
-		klog.Errorf("Error converting distrNpcs: %v", err)
 		return nil, err
 	}
 
-	var compression bool
-	var encryption bool
+	compression := getBoolParameter(params, "compression")
+	encryption := getBoolParameter(params, "encryption")
```
Comment (boddumanohar) [inline pkg/spdk/controllerserver.go:230]: nice work moving them to separate functions

File: pkg/spdk/controllerserver.go [hunk 2]
```diff
@@ -194,59 +198,69 @@ func (cs *controllerServer) DeleteSnapshot(_ context.Context, req *csi.DeleteSna
...
 
-	if req.GetParameters()["compression"] == "true" ||
-		req.GetParameters()["compression"] == "True" {
-		compression = true
-	}
+	pvcName, pvcNameSelected := params["csi.storage.k8s.io/pvc/name"]
```
Comment (boddumanohar) [inline pkg/spdk/controllerserver.go:232]: can you please create a constant: `csi.storage.k8s.io/pvc` and prepare the parameter key by appending `name` or `namespace to it.

File: pkg/spdk/controllerserver.go (hunk 1 above)
Comment (boddumanohar) [inline pkg/spdk/controllerserver.go:230]: we need to fail volume creation if encryption flag is set but the keys are not provided.

File: pkg/spdk/controllerserver.go [hunk 3]
```diff
@@ -539,3 +553,39 @@ func newControllerServer(d *csicommon.CSIDriver) (*controllerServer, error) {
 
 	// return &server, nil
 }
+
+func GetCryptoKeys(ctx context.Context, pvcName, pvcNamespace string) (cryptoKey1, cryptoKey2 string, err error) {
```
Comment (boddumanohar) [inline pkg/spdk/controllerserver.go:557]: we might need to add a clusterRole to get secrets from the namespace.

**Review by boddumanohar — CHANGES_REQUESTED** [review]

File: pkg/spdk/controllerserver.go (hunk 1 above)
Comment (geoffrey1330) [reply pkg/spdk/controllerserver.go:230]: Alright.

File: pkg/spdk/controllerserver.go (hunk 2 above)
Comment (geoffrey1330) [reply pkg/spdk/controllerserver.go:232]: Sure

File: pkg/spdk/controllerserver.go (hunk 3 above)
Comment (geoffrey1330) [reply pkg/spdk/controllerserver.go:569]: A clusterrole already exist [here](https://github.com/simplyblock-io/spdk-csi/blob/master/charts/spdk-csi/templates/controller-rbac.yaml#L24)

File: charts/spdk-csi/templates/controller.yaml [hunk 4]
```diff
@@ -31,6 +31,7 @@ spec:
         - "--timeout=30s"
         - "--retry-interval-start=500ms"
         - "--leader-election=false"
+        - "--extra-create-metadata=true"
```
Comment (boddumanohar) [inline charts/spdk-csi/templates/controller.yaml:34]: could you please remind me why was this additional flag added?

**Review by boddumanohar — COMMENTED** [review]
rest of the changes looks good. can you please add docs in `docs` folder discussion how about how to generate the key. 

You could take reference from this https://github.com/simplyblock-io/sbcli/blob/main/docs/create_crypto_lvol.md

File: charts/spdk-csi/templates/controller.yaml (hunk 4 above)
Comment (geoffrey1330) [reply charts/spdk-csi/templates/controller.yaml:34]: This field was added so we could get the metadata pvc name and namespace. so we could be able to call it via `request.Getparamter()`

**Comment by geoffrey1330** [comment]
> rest of the changes looks good. can you please add docs in `docs` folder discussion how about how to generate the key.
> 
> You could take reference from this https://github.com/simplyblock-io/sbcli/blob/main/docs/create_crypto_lvol.md

Alright will do that

File: charts/spdk-csi/templates/controller.yaml (hunk 4 above)
Comment (boddumanohar) [reply charts/spdk-csi/templates/controller.yaml:34]: Good find. Thanks for clarifying. Great work 💯

**Review by boddumanohar — APPROVED** [review]

**Review by boddumanohar — APPROVED** [review]
awesome 🚀

**Pull request merged**